libass/ass.c

```diff
@@ -186,6 +191,32 @@ static int resize_read_order_bitmap(ASS_Track *track, int max_id)
     return -1;
 }
 
+static int compare_event_deadline(const void *a, const void *b)
+{
+    const ASS_Event *ev1 = (const ASS_Event*)a;
+    const ASS_Event *ev2 = (const ASS_Event*)b;
+    long long diff = (ev1->Start - ev2->Start) + (ev1->Duration - ev2->Duration);
+    if (diff > 0) return -1;
+    if (diff < 0) return 1;
+    return 0;
+}
+
+// This function is not thread safe
+int ass_prune_history(ASS_Track *track, long long deadline, int size_threshold)
+{
+    int n_events_ini = track->n_events;
+    if (size_threshold >= 0 && (int64_t)size_threshold << 10 < track->parser_priv->events_size) {
+        //undisplay time descending order
+        qsort((void*)track->events, track->n_events, sizeof(ASS_Event), compare_event_deadline);
+        ASS_Event *ev = track->events + track->n_events - 1;
+        while (track->n_events > 0 && ev->Start + ev->Duration < deadline) {
+            ass_free_event(track, --track->n_events);
+            --ev;
+        }
+    }
+    return n_events_ini-track->n_events;
+}
```

Comment:
There are several issues with this sorting:

 - `qsort` isn’t (guaranteed to be) a stable sort, yet rendering depends on the order of events, introducing random display differences
 - even a stable sort by end time might mix this order up in a detectable way if two events start simultaneously but the originally later-defined event end earlier
 - *(sorting by start time should be side-effect-free for a continuous watch, but this doesn't help much with this cleanup and i’m not fully sure if there might be visible differences when skipping around or if we should even care about the latter)*

Reply:
I could sort considering both the event id and undisplay timestamp... or just go for thepiecewise memmove or memcpy to ensure the event order is strictly preserved.